### DISCUSSION TITLE
RoadWarrior IPSec with GRE scenario is not doing ESP encryption

### DISCUSSION THREAD

**Discussion by dhanaseelangit** [strongswan/strongswan#1696] [General]
I am using the below topology for making RoadWarrior GRE case between router moon(referring as server for ipsec connection) and device carol,
 
10.1.0.0/16 -- | 192.168.0.1 | === | 192.168.0.100
  moon-net          moon                 carol (Virtual IP [10.3.0.1)]
 
Swanctl at server side :
      children {
        net {
            # For GRE
            local_ts  = dynamic[gre]
            remote_ts = dynamic[gre]
            #mode = transport
 
            #unique = keep
            #updown = /usr/local/libexec/ipsec/_updown iptables
            esp_proposals = aes128gcm128-x25519
        }
 
Have configured GRE tunnel at host as below : 
            ip tunnel add gre-moon local 172.17.0.4 remote 172.17.0.5 mode gre key 42
 
 
Also added the explicit route :
            ip route add 10.3.0.0/24 dev gre-moon
 
Swanctl at Host side:
 
      children {
        home {
           local_ts  = dynamic[gre]
           remote_ts = dynamic[gre]
           #mode = transport
           esp_proposals = aes128gcm128-x25519
        }
 
Have configured GRE tunnel at host as below : 
            ip link add name gre-carol type gre key 42 ttl 64 remote 172.17.0.4 dev eth1
            ip link set mtu 1350 dev gre-carol
            ip link set multicast on dev gre-carol
            ip link set up dev gre-carol
        ip addr add 10.3.0.1/32 dev gre-carol
 
 
Strongswan.conf is configured with install_routes = 0 on both end.
 
 
I could able to establish a IPSEC tunnel between Server and Host with GRE tunnel. Below logs are from Host side,
 
root@d1375b90c412:/# swanctl --list-sas
home: #1, ESTABLISHED, IKEv2, 3128331c8c9e8797_i* 27988fa86545e43b_r
  local  'carol@strongswan.org' @ 172.17.0.5[4500]
  remote 'moon.strongswan.org' @ 172.17.0.4[4500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519
  established 1484s ago, rekeying in 12561s
  home: #1, reqid 1, INSTALLED, TUNNEL, ESP:AES_GCM_16-128
    installed 1484s ago, rekeying in 1998s, expires in 2476s
    in  c2b2bda2,   4928 bytes,    44 packets,  1305s ago
    out c63640e4,   4928 bytes,    44 packets,  1307s ago
    local  172.17.0.5/32[gre]
    remote 172.17.0.4/32[gre]
root@d1375b90c412:/# 
 
But I could not able to route any of the packets via the GRE/IPSEC tunnels. As per the strongswan page, 
 
root@d1375b90c412:/#  ip -s  tunnel show 
gre0: gre/ip remote any local any ttl inherit nopmtudisc
RX: Packets    Bytes        Errors CsumErrs OutOfSeq Mcasts
    0          0            0      0        0        0       
TX: Packets    Bytes        Errors DeadLoop NoRoute  NoBufs
    0          0            0      0        0        0     
gre-carol: gre/ip remote 172.17.0.4 local any dev eth0 ttl 64 key 42
RX: Packets    Bytes        Errors CsumErrs OutOfSeq Mcasts
    2077       174468       0      0        0        0       
TX: Packets    Bytes        Errors DeadLoop NoRoute  NoBufs
    2154       180936       0      0        0        0     
root@d1375b90c412:/# 
 
root@d1375b90c412:/# ip xfrm p
src 172.17.0.5/32 dst 172.17.0.4/32 proto gre 
	dir out priority 366975 ptype main 
	tmpl src 172.17.0.5 dst 172.17.0.4
		proto esp spi 0xc63640e4 reqid 1 mode tunnel
src 172.17.0.4/32 dst 172.17.0.5/32 proto gre 
	dir fwd priority 366975 ptype main 
	tmpl src 172.17.0.4 dst 172.17.0.5
		proto esp reqid 1 mode tunnel
src 172.17.0.4/32 dst 172.17.0.5/32 proto gre 
	dir in priority 366975 ptype main 
	tmpl src 172.17.0.4 dst 172.17.0.5
		proto esp reqid 1 mode tunnel
 
root@d1375b90c412:/# ip xfrm s
src 172.17.0.5 dst 172.17.0.4
	proto esp spi 0xc63640e4 reqid 1 mode tunnel
	replay-window 0 flag af-unspec
	aead rfc4106(gcm(aes)) 0x3ddf8bbe51716d6a2289a8c5342466ae138050ee 128
	anti-replay context: seq 0x0, oseq 0x2c, bitmap 0x00000000
src 172.17.0.4 dst 172.17.0.5
	proto esp spi 0xc2b2bda2 reqid 1 mode tunnel
	replay-window 32 flag af-unspec
	aead rfc4106(gcm(aes)) 0xe80da90b252e470ba06c103fd12aeb56c4eed580 128
	anti-replay context: seq 0x2c, oseq 0x0, bitmap 0xffffffff
root@d1375b90c412:/# 
 
root@d1375b90c412:/#  iptables --list
# Warning: iptables-legacy tables present, use iptables-legacy to see them
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere            

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
root@d1375b90c412:/# 
 
Can you please help me to understand below points?
	1. Can you please confirm the **GRE IPSec is valid for Remote Warrior case. Any such test done or setup tested earlier?** Since there is no reference KVM test logs in strongswan page(https://www.strongswan.org/testing/testresults/route-based/index.html), **I tried framing the RW configs based to net2netGRE configuration**.
        2. If remote warrior is supported, Can you please help me to understand how this configurations can be corrected to route the traffic with ESP encryption successfully in both direction. **As per xfrm policy table, the tunnel is established between 172.17.0.5 and 172.17.0.4 and the Traffic selector is set as 172.17.0.5  and 172.17.0.4 in **IN and OUT and FWD policy**. I am **able to send the traffic from Host to Server, but packet capture shows only the ICMP packet not an ESP encrypted packet**.
        3. I have framed the ip tunnel config based on the updown script given in the strongswan page. I had seen **updown plugin** available in strongswan. Will that be helpful here. As per strongswan page “ To accept traffic with default DROP policies, one may alternatively use global, non-tunnel specific rules matching IPsec traffic with the Netfilter policy match.”. Can you please help me to understand this.
        4. Also **can you please confirm the load tester plugin supports remote-warrior GRE**?
 
Thanks,
Dhana

**Comment by tobiasbrunner** [comment]
What exactly makes this a road warrior scenario? Also, where do the 172.17.* addresses come from, they are not in your topology diagram. Anyway, you haven't specified the local address of the GRE interface on "carol". I don't think this works without it (no idea what the source IP of the GRE packet would even be).

> 3. I have framed the ip tunnel config based on the updown script given in the strongswan page. I had seen **updown plugin** available in strongswan. Will that be helpful here. As per strongswan page “ To accept traffic with default DROP policies, one may alternatively use global, non-tunnel specific rules matching IPsec traffic with the Netfilter policy match.”. Can you please help me to understand this.

The _updown_ plugin calls the configured script. The script can do all sorts of things and something the default script does is install firewall rules that allow traffic if the default policies for e.g. INPUT and FORWARD are DROP/REJECT. But as documented, such traffic can also be allowed with more generic rules that are not installed dynamically by a script.

> 4. Also **can you please confirm the load tester plugin supports remote-warrior GRE**?

It doesn't support any type of route-based setups.

**Reply by dhanaseelangit** [reply]
Hi Tobias,

Thanks a lot for the reply and the clarifications. 

_What exactly makes this a road warrior scenario? Also, where do the 172.17.* addresses come from, they are not in your topology diagram. Anyway, you haven't specified the local address of the GRE interface on "carol". I don't think this works without it (no idea what the source IP of the GRE packet would even be)._

As per my understanding the GRE tunnel requires endpoints at both ends. Here, Moon(172.17.0.4) and Carol(172.17.0.5) are the end points for the GRE tunnel and the **mode I chosen as Tunnel instead of Transport**. I don't see much difference between the config of RW case and net2net case as per my understanding. As per my config, the virtual ip differentiates the client as Carol. Please correct me if my understanding is wrong.

I am using the ping command "ping 10.1.0.1 -I gre-carol -c 5" to send a traffic to moon server. 10.1.0.1 is the address of local interface on the moon. With this ping command, the traffic is pushed via the GRE tunnel at Carol side. But, since the ip xframe policy "src 172.17.0.5 dst 172.17.0.4" is not matching with the packet, ESP encryption is not happening which leads to transfer plain ICMP packet on the tunnel.

I have configure the local_ts and remote_ts as dynamic in my config but the system is updating the "src 172.17.0.5 dst 172.17.0.4" as my local and remote_ts for the policy. Please let me know if I am missing anything here. Attached the full swanctl config file of both server and client.

root@d1375b90c412:/# ip xfrm p
src 172.17.0.5/32 dst 172.17.0.4/32 proto gre
    dir out priority 366975 ptype main
    **tmpl src 172.17.0.5 dst 172.17.0.4
        proto esp spi 0xc63640e4 reqid 1 mode tunnel**
src 172.17.0.4/32 dst 172.17.0.5/32 proto gre
    dir fwd priority 366975 ptype main
    **tmpl src 172.17.0.4 dst 172.17.0.5
        proto esp reqid 1 mode tunnel**
src 172.17.0.4/32 dst 172.17.0.5/32 proto gre
    dir in priority 366975 ptype main
    **tmpl src 172.17.0.4 dst 172.17.0.5
        proto esp reqid 1 mode tunnel**

Swanctl config of Carol(Client) :

connections {
    home {
          local_addrs = 172.17.0.5
          remote_addrs = 172.17.0.4
          local {
              auth = pubkey
              certs = carolCert.pem
              id = carol@strongswan.org
            }
          remote {
            auth = pubkey
            id = moon.strongswan.org
          }
          children {
                  home {
                     local_ts  = dynamic[gre]
                     remote_ts = dynamic[gre]
                     esp_proposals = aes128gcm128-x25519
                  }
                  version = 2
                  proposals = aes128-sha256-x25519
          }
    }
}

secrets {
    rsa {
       file = initiator_key.pem
    }

Swanctl config of Server (moon): 

  connections {
        rw {
          local_addrs  = 172.17.0.4
          local {
            auth = pubkey
            certs = moonCert.pem
            id = moon.strongswan.org
          }
          remote {
            auth = pubkey
          }
          children {
            net {
                    # For GRE
                    local_ts  = dynamic[gre]
                    remote_ts = dynamic[gre]
                    #mode = transport 
                    esp_proposals = aes128gcm128-x25519
            }
          }
          version = 2
          proposals = aes128-sha256-x25519
    }
  }

  pools {
      rw_pool {
         addrs = 10.3.0.0/16 
      }
  }
Thanks,
Dhana

**Reply by tobiasbrunner** [reply]
> As per my understanding the GRE tunnel requires endpoints at both ends.

Yes, but you haven't configured the local endpoint on the GRE interface on carol with the command you gave in your original post.







> the **mode I chosen as Tunnel instead of Transport**.

That doesn't really matter. It just adds additional overhead if the inner and outer IPs are the same.





> I don't see much difference between the config of RW case and net2net case as per my understanding. As per my config, the virtual ip differentiates the client as Carol.

Where does any virtual IP come into play? I don't see that in your config. You just create a host-to-host tunnel that's restricted to the GRE protocol.



> But, since the ip xframe policy "src 172.17.0.5 dst 172.17.0.4" is not matching with the packet, ESP encryption is not happening which leads to transfer plain ICMP packet on the tunnel.

If it's a plain ICMP packet, then GRE encapsulation already doesn't work. Maybe you are missing a route that goes via GRE interface (and again, that interface might not be configured correctly without local address).

**Reply by dhanaseelangit** [reply]
Hi Tobias,

Thanks once again for the reply.

_As per my understanding the GRE tunnel requires endpoints at both ends.

Yes, but you haven't configured the local endpoint on the GRE interface on carol with the command you gave in your original post._

**Below is the GRE tunnel config as per my original post,
    ip link add name gre-carol type gre key 42 ttl 64 remote 172.17.0.4 dev eth1
    ip link set mtu 1350 dev gre-carol
    ip link set multicast on dev gre-carol
    ip link set up dev gre-carol
    ip addr add 10.3.0.1/32 dev gre-carol**

The last line here adds the local end point address as 10.3.0.1 (This is the virtual ip address assigned for the Client Carol) for the GRE tunnel.

My scenario is like, I will be creating multiple RW VPN clients, each should be associated with its own GRE tunnel endpoint. 

_I don't see much difference between the config of RW case and net2net case as per my understanding. As per my config, the virtual ip differentiates the client as Carol.

Where does any virtual IP come into play? I don't see that in your config. You just create a host-to-host tunnel that's restricted to the GRE protocol._

**While copy&pasting the configuration here, i removed the commented lines. Seems i removed the vips config as part of that. I added the config "vips = 0.0.0.0" under client swanctl.config file,
          connections {
              home {
                local_addrs = 172.17.0.5
                remote_addrs = 172.17.0.4
                vips = 0.0.0.0
                local {**

But, since the ip xframe policy "src 172.17.0.5 dst 172.17.0.4" is not matching with the packet, ESP encryption is not happening which leads to transfer plain ICMP packet on the tunnel.

If it's a plain ICMP packet, then GRE encapsulation already doesn't work. Maybe you are missing a route that goes via GRE interface (and again, that interface might not be configured correctly without local address).

Seems the ICMP is encapsulated in with GRE header and the same is visible in wireshark pcap (Link layer address type : GRE over IP (778)) under Linux cooked capture v1.
![16842472546907307468300773864869](https://github.com/strongswan/strongswan/assets/122854660/570de8e5-a225-4828-a06e-8667748b5546)

Thanks,
Dhana

**Reply by tobiasbrunner** [reply]
> The last line here adds the local end point address as 10.3.0.1

No, that's just an IP on the interface, it's not the local address of the GRE tunnel (which would be set via `local` parameter for the `ip link add` call). The address you added could serve as source IP for routed traffic, but it won't be the source address for the GRE packets.

**Comment by dhanaseelangit** [comment]
Hi Tobias,

So it means the ip address 172.17.0.5 should be the local address for  both
IPsec tunnel and GRE tunnel at carol..similarly 172.17.0.4 should be for
both IPsec and GRE at server??

If that's the case, do I need to create separate interface for each IPsec
VPN clients I am trying to create??

Thanks,
Dhana

**Reply by tobiasbrunner** [reply]
The endpoints of the GRE tunnel have to match the traffic selector, that could but doesn't have to match the endpoints of the IKE and ESP tunnel (if you use tunnel mode, in transport mode they have to match). So you could use a virtual IP in the traffic selector, but then you have to create the GRE interface on client and server accordingly so the endpoints match.

**Reply by dhanaseelangit** [reply]
Hi Tobias,

Sorry for the late reply. I was bit busy with the other works and I could not able to concentrate on this for a while..

Thanks a lot for the reply.. I tried the same suggested by you, below are the observation i had,

On a Roadwarrior Case with Virtual IP, I am configuring the local_ts and remote_ts as dynamic[gre] and i could see the "ip xfrm policy" is loaded with traffic selector as 172.17.0.9(local endpoint of carol used as local_addr in IKE_SA config) and 172.17.0.8 (Endpoint of moon's address used as remote_addr in IKE_SA config). 

root@b27e774e04d3:/# ip xfrm p
src 10.3.0.1/32 dst 172.17.0.8/32 proto gre 
	dir out priority 366975 ptype main 
	**tmpl src 172.17.0.9 dst 172.17.0.8**
		proto esp spi 0xc221fbab reqid 1 mode tunnel
src 172.17.0.8/32 dst 10.3.0.1/32 proto gre 
	dir fwd priority 366975 ptype main 
	**tmpl src 172.17.0.8 dst 172.17.0.9**
		proto esp reqid 1 mode tunnel
src 172.17.0.8/32 dst 10.3.0.1/32 proto gre 
	dir in priority 366975 ptype main 
	**tmpl src 172.17.0.8 dst 172.17.0.9**
		proto esp reqid 1 mode tunnel

Which expects the traffic with src 172.17.0.9 and dst 172.17.0.8 only be allowed on the GRE/IPSec tunnel. So the traffic I am passing with destination 10.1.0.1(network reachable on moon) is not reaching. Do i need to add any specific config apart from ip route to make it work?

The same I tried with mode transport also and its not working in that case too.

The only case working for me is with Roadwarrior without Virtual IP and configured with mode as Transport, the "ip xfrm policy" below which clearly has src and dst of traffic selectors as 0.0.0.0(any) which works properly.

root@d1375b90c412:/# ip xfrm p
src 172.17.0.5/32 dst 172.17.0.4/32 proto gre 
	dir out priority 366975 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp spi 0xc44678dd reqid 1 mode transport
src 172.17.0.4/32 dst 172.17.0.5/32 proto gre 
	dir in priority 366975 ptype main 
	tmpl src 0.0.0.0 dst 0.0.0.0
		proto esp reqid 1 mode transport

Can you please help in understanding/correcting the configuration for Roadwarrior with Virtual IP case?

Thanks,
Dhana

**Reply by tobiasbrunner** [reply]
You have that wrong, the first line is the selector of the respective policy, the two lines that start with `tmpl` are the link to the IPsec SA. So in the first scenario the TS are the virtual IP (`10.3.0.1/32`) and the peer's public IP (`172.17.0.8/32`), that is, the GRE tunnel endpoints would have to be these addresses to match the policies.  In the second scenario, you directly use the two peer's physical IPs as TS (nothing else is possible in transport mode), so the endpoints of the GRE tunnel are these two addresses. In both cases traffic from/to any IP can be sent through the GRE tunnel.

> Can you please help in understanding/correcting the configuration for Roadwarrior with Virtual IP case?

I still don't understand what exactly you want to achieve. How exactly is the virtual IP to be used? As endpoint of the GRE tunnel? Or as source IP for the traffic that's tunneled through the GRE tunnel?

**Reply by dhanaseelangit** [reply]
Hi Tobias,

Thanks for the reply.

_Can you please help in understanding/correcting the configuration for Roadwarrior with Virtual IP case?

I still don't understand what exactly you want to achieve. How exactly is the virtual IP to be used? As endpoint of the GRE tunnel? Or as source IP for the traffic that's tunneled through the GRE tunnel?_
[Dhana] I am trying to make Virtual IP as TS/GRE tunnel end point for carol. So the traffic sent from the Virtual IP should be sent via GRE tunnel and it should be encrypted.

Thanks,
Dhana

**Reply by tobiasbrunner** [reply]
> I am trying to make Virtual IP as TS/GRE tunnel end point for carol. So the traffic sent from the Virtual IP should be sent via GRE tunnel and it should be encrypted.

Again, this doesn't clarify things really. "make Virtual IP as TS/GRE tunnel endpoint" isn't the same thing as "traffic sent from the virtual IP should be sent via GRE tunnel", that's exactly what I already pointed out above. For the former you need to negotiate virtual IPs with the peer so your local TS is the VIP, but then the GRE interface has to use matching endpoints. For the latter, you don't need to negotiate virtual IPs but instead negotiate a host-to-host tunnel, and then just install the IP you want to use inside the tunnel on the GRE interface (and possibly reference it as preferred source IP in the route via that interface).

**Reply by dhanaseelangit** [reply]
Hi Tobias,

Yes I got the point clearly now and able to make the setup and send the traffic..

08:15:54.911417 IP 172.17.0.8.500 > 172.17.0.9.500: isakmp: child_sa  inf2[IR]
08:16:33.949067 IP 172.17.0.8 > 172.17.0.9: ESP(spi=0xc314916e,seq=0x1), length 100
08:16:33.949135 IP 10.1.0.1 > 10.3.0.2: GREv0, key=0x2b, length 36: IP 10.1.0.2 > 10.3.0.2: ICMP echo request, id 54280, seq 0, length 8
08:16:33.949177 IP 10.3.0.1 > 10.1.0.1: GREv0, key=0x2a, length 92: IP 10.3.0.2 > 10.1.0.1: ICMP 10.3.0.2 protocol 47 port 2048 unreachable, length 64
08:16:33.949213 IP 172.17.0.9 > 172.17.0.8: ESP(spi=0xce2161fa,seq=0x1), length 164
08:16:33.949245 IP 172.17.0.8 > 172.17.0.9: ESP(spi=0xc314916e,seq=0x2), length 100
08:16:33.949277 IP 10.1.0.1 > 10.3.0.2: GREv0, key=0x2b, length 36: IP 10.1.0.2 > 10.3.0.2: ICMP echo request, id 54280, seq 256, length 8
08:16:33.949312 IP 10.3.0.1 > 10.1.0.1: GREv0, key=0x2a, length 92: IP 10.3.0.2 > 10.1.0.1: ICMP 10.3.0.2 protocol 47 port 2048 unreachable, length 64
08:16:33.949347 IP 172.17.0.9 > 172.17.0.8: ESP(spi=0xce2161fa,seq=0x2), length 164

Thank a lot four your kind response.

Thanks,
Dhana
